receiver/k8sclusterreceiver/watcher.go

```diff
@@ -85,8 +95,21 @@ func (rw *resourceWatcher) prepareSharedInformerFactory() {
 }
 
 // startWatchingResources starts up all informers.
-func (rw *resourceWatcher) startWatchingResources(stopper <-chan struct{}) {
-	rw.sharedInformerFactory.Start(stopper)
+func (rw *resourceWatcher) startWatchingResources(ctx context.Context) {
+	var cancel context.CancelFunc
+	rw.timedContextForInitialSync, cancel = context.WithTimeout(ctx, rw.initialTimeout)
+
+	// Start off individual informers in the factory.
+	rw.sharedInformerFactory.Start(ctx.Done())
+
+	// Ensure cache is synced with initial state, once informers are started up.
+	// Note that the event handler can start receiving events as soon as the informers
+	// are started. So it's required to ensure that the receiver does not start
+	// collecting data before the cache sync since all data may not be available.
+	// This method will block either till the timeout set on the context, until
+	// the initial sync is complete or the parent context is cancelled.
+	rw.sharedInformerFactory.WaitForCacheSync(rw.timedContextForInitialSync.Done())
+	cancel()
```

Comment:
would be `defer cancel()` better instead?

Reply:
In this case since it's the last line to be executed within the scope of the method, I believe it shouldn't make a difference.

Reply:
no difference, just for consistency with general practice. anyway your call

Reply:
👍 updated.